_posts/2018-10-05-a-macroscopic-take-on-metaprogramming.md

```diff
@@ -0,0 +1,293 @@
+Now we leave all properties but one alone and focus entirely on 'trapping' `_getDispatcher`.
+When it's called, we output the arguments to the console.
+
+It's a different, slightly more concise more way of implementing our solution.
+
+## Elixir
+
+In Elixir, most of the metaprogramming boils down to understanding the functions "quote" and "unquote" (by no means a measure of simplicity).
+Quote effectively gives us access to Elixir's AST (abstract syntax tree) by showing us how an expression is parsed and transformed into a tree that Elixir can handle.
+What is returned typically contains three-element tuples (operation/data type, metadata, arguments/data).
+
+### Example
+
+Again, we would like to make our test suite a bit more concise.
+Our application offers two endpoints with a very similar set of functionality.
+This should be a good case for abstracting most of their testing and sharing the tests between the two endpoints (endpoint_a and endpoint_b).
+Not dissimilar to RSpec shared example feature, making use of a domain specific language would work for our purposes.
+
+We have a series of tests like the one below:
+
+```
+defmodule Api.V1.AControllerTest do
+	test "#create http returns a 403 status with json error" do
+	  "/v1/endpoint_a"
+	  |> create_record(403, ssl_error())
+	end
+end
+
+defmodule Api.V1.BControllerTest do
+	test "#create http returns a 403 status with json error" do
+	  "/v1/endpoint_b"
+	  |> create_record(403, ssl_error())
+	end
+end
+
+```
+
+The only difference being the endpoint, we can abstract this to:
+
+```
+defmodule SharedTestCase do
+  use ExUnit.Case
+
+  @moduletag endpoint: "endpoint_a"
+
+  test "#create http returns a 403 status with json error" do
+  		"/v1/#{endpoint}"
+  		|> create_record(403, ssl_error())
+	end
+end
+```
+
+@moduletag is a module attribute provided by ExUnit: one of its uses being to act as a temporary module storage to be used during compilation.
+
+A macro is going to help Api.V1.AControllerTest and Api.V1.BControllerTest modules make use of the SharedTestCase module.
+
+```
+defmodule SharedTestCase do
+  defmacro __using__(options) do
+    quote do
+      use ExUnit.Case
+
+      @moduletag unquote(options)
+
+      test "#create http returns a 403 status with json error", %{endpoint: endpoint} do
+        "/v1/#{endpoint}"
+        |> create_record(403, ssl_error())
+      end
+
+      ...
+    end
+  end
+end
+```
+
+It's possible to go even further by extracting all the tests to another macro to define the tests separately.
+
+
+Metaprogramming is not a bag of tricks, it's the path to a deeper understanding of any language and potentially more expressive code.
+
+Languages like Elixir give the ability to the programmer of making extensions tailored to their needs which are first-class features.
+
+It's too good an opportunity to pass up and not use for the greater good of generalisation, DRY-ing your codebase and making the code more expressive.
```

Comment:
generalisation => Z!

Reply:
Apparently I did not push that particular change! I swear I had caught it 😬